ldap/servers/slapd/log.c

```diff
@@ -3013,137 +3204,969 @@ log__open_accesslogfile(int logfile_state, int locked)
+                      "Moving to a new security log file (%s)\n",
+                      pathname, 0, 0);
+
+        LOG_CLOSE(loginfo.log_security_fdes);
+        loginfo.log_security_fdes = 0;
+        loginfo.log_security_ctime = 0;
+        logp = loginfo.log_security_logchain;
+        while (logp) {
+            d_logp = logp;
+            logp = logp->l_next;
+            slapi_ch_free((void **)&d_logp);
+        }
+        loginfo.log_security_logchain = NULL;
+        slapi_ch_free_string(&loginfo.log_security_file);
+        loginfo.log_security_file = NULL;
+        loginfo.log_numof_security_logs = 1;
+    }
+
+    /* Now open the new security log file */
+    if (security_log_openf(pathname, 1 /* locked */)) {
+        rv = LDAP_LOCAL_ERROR; /* error Unable to use the new dir */
+    }
+    LOG_SECURITY_UNLOCK_WRITE();
+    return rv;
+}
+
+/******************************************************************************
+* security_log_openf
+*
+*     Open the security log file
+*
+*  Returns:
+*    0 -- success
+*    1 -- fail
+******************************************************************************/
+int
+security_log_openf(char *pathname, int locked)
+{
+    int rv = 0;
+    int logfile_type = 0;
+
+    if (!locked)
+        LOG_SECURITY_LOCK_WRITE();
+
+    /* store the path name */
+    slapi_ch_free_string(&loginfo.log_security_file);
+    loginfo.log_security_file = slapi_ch_strdup(pathname);
+
+    /* store the rotation info fiel path name */
+    slapi_ch_free_string(&loginfo.log_securityinfo_file);
+    loginfo.log_securityinfo_file = slapi_ch_smprintf("%s.rotationinfo", pathname);
+
+    /*
+     * Check if we have a log file already. If we have it then we need to parse
+     * the header info and update the loginfo struct.
+     */
+    logfile_type = log__security_rotationinfof(loginfo.log_securityinfo_file);
+
+    if (log__open_securitylogfile(logfile_type, 1 /* got lock*/) != LOG_SUCCESS) {
+        rv = 1;
+    }
+
+    if (!locked)
+        LOG_SECURITY_UNLOCK_WRITE();
+
+    return rv;
+}
+
+static void
+log_append_security_buffer(time_t tnl, LogBufferInfo *lbi, char *msg, size_t size)
+{
+    slapdFrontendConfig_t *slapdFrontendConfig = getFrontendConfig();
+    char *insert_point = NULL;
+
+    /* While holding the lock, we determine if there is space in the buffer for our payload,
+       and if we need to flush.
+     */
+    PR_Lock(lbi->lock);
+    if (((lbi->current - lbi->top) + size > lbi->maxsize) ||
+        (tnl >= loginfo.log_security_rotationsyncclock &&
+         loginfo.log_security_rotationsync_enabled))
+    {
+        log_flush_buffer(lbi, SLAPD_SECURITY_LOG, 0 /* do not sync to disk */);
+    }
+    insert_point = lbi->current;
+    lbi->current += size;
+    /* Increment the copy refcount */
+    slapi_atomic_incr_64(&(lbi->refcount), __ATOMIC_RELEASE);
+    PR_Unlock(lbi->lock);
+
+    /* Now we can copy without holding the lock */
+    memcpy(insert_point, msg, size);
```

Comment:
Just curious, is there a risk that the log file get rotated between unlock and memcpy so that insert_point is invalid ?

Reply:
@tbordaz I'd have to check, but this is how the old code did it.  I just renamed the function, so if there is a problem, it has always been there.